src/sentry/lang/native/symbolicator.py

```diff
@@ -0,0 +1,154 @@
+from __future__ import absolute_import
+
+import six
+
+import time
+import logging
+
+from django.core.urlresolvers import reverse
+
+from requests.exceptions import RequestException
+
+from sentry import options
+from sentry.cache import default_cache
+from sentry.utils import metrics
+from sentry.net.http import Session
+from sentry.tasks.store import RetrySymbolication
+
+MAX_ATTEMPTS = 3
+REQUEST_CACHE_TIMEOUT = 3600
+SYMBOLICATOR_TIMEOUT = 5
+
+logger = logging.getLogger(__name__)
+
+
+def run_symbolicator(stacktraces, modules, project, arch, signal, request_id_cache_key):
+    self_url_prefix = options.get('system.internal-url-prefix') \
+        or options.get('system.url-prefix')
+
+    assert self_url_prefix
+    self_bucket_url = '%s%s' % (
+        self_url_prefix.rstrip('/'),
+        reverse('sentry-api-0-dsym-files', kwargs={
+            'organization_slug': project.organization.slug,
+            'project_slug': project.slug
+        })
+    )
+
+    project_id = six.text_type(project.id)
+    request_id = default_cache.get(request_id_cache_key)
+    sess = Session()
+
+    attempts = 0
+    wait = 0.5
+
+    with sess:
+        while 1:
+            try:
+                rv = _do_send_request(
+                    sess=sess,
+                    request_id=request_id,
+                    project_id=project_id,
+                    self_bucket_url=self_bucket_url,
+                    signal=signal,
+                    stacktraces=stacktraces,
+                    modules=modules
+                )
+
+                metrics.incr('events.symbolicator.status.%s' % rv.status_code, tags={
+                    'project_id': project_id
+                })
+
+                if rv.status_code == 404 and request_id:
+                    default_cache.delete(request_id_cache_key)
+                    request_id = None
+                    continue
+                elif rv.status_code == 503:
+                    raise RetrySymbolication(retry_after=10)
+
+                rv.raise_for_status()
```

Comment:
Putting this up for discussion: What if we don't raise here but record a fatal (non-solvable) processing error here? That way, at least the event is not dropped and people see something in the issues stream. This might be easier to tackle for support requests than events that simply do not show up.

Reply:
The events are not dropped. This is a `RequestException` which gets caught in the try-except around the entire while-loop-body.

FWIW this part is identical to how requests are send to the iOS symbolserver.

Reply:
Completely missed that, okay that's perfect then. Let's add a processing error then, to signal this to the user. 

I know that this is not consistent with the iOS symbol server, which only logs an error. However, since this time user symbols are involved, I want to make it clear in the UI that a failure of our symbolicator is not the user's fault by showing some explicit message.

Reply:
Processing error is still missing.

Reply:
Will do!

Reply:
@jan-auer I wonder if we should just silently fall back to symbolic when this fails for now?

Reply:
Hm, let's stick with an error. This might help us connect issues to customer reports, and also prevents doing work twice.

Reply:
done